File: sdks/go/rpc/endpoint_test.go
package rpc

import (
	"errors"
	"strings"
	"testing"
)

// Unit tests for ParseEndpoint.
//
// The contract is shared with the Rust + Python + TypeScript SDKs;
// the golden URL forms here are the same ones tested in
// sdks/rust/src/rpc/endpoint.rs, sdks/python/tests/rpc/test_endpoint.py,
// and sdks/typescript/tests/rpc/endpoint.test.ts. Keeping them aligned
// across SDKs prevents subtle behaviour drift between language runtimes.

func TestParseEndpoint_RPCWithExplicitHostAndPort(t *testing.T) {
	ep, err := ParseEndpoint("vectorizer://example.com:9000")
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if ep.Kind != EndpointRPC || ep.Host != "example.com" || ep.Port != 9000 {
		t.Fatalf("got %+v", ep)
	}
}

func TestParseEndpoint_RPCWithoutPortDefaultsTo15503(t *testing.T) {
	ep, err := ParseEndpoint("vectorizer://example.com")
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if ep.Kind != EndpointRPC || ep.Host != "example.com" || ep.Port != DefaultRPCPort {
		t.Fatalf("got %+v", ep)
	}
	if DefaultRPCPort != 15503 {
		t.Fatalf("DefaultRPCPort drifted: %d", DefaultRPCPort)
	}
}

func TestParseEndpoint_BareHostPortIsRPC(t *testing.T) {
	ep, err := ParseEndpoint("localhost:15503")
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if ep.Kind != EndpointRPC || ep.Host != "localhost" || ep.Port != 15503 {
		t.Fatalf("got %+v", ep)
	}
}

func TestParseEndpoint_HTTPRoutesToREST(t *testing.T) {
	ep, err := ParseEndpoint("http://localhost:15002")
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if ep.Kind != EndpointREST || ep.URL != "http://localhost:15002" {
		t.Fatalf("got %+v", ep)
	}
	ep, err = ParseEndpoint("https://api.example.com")
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if ep.Kind != EndpointREST || ep.URL != "https://api.example.com" {
		t.Fatalf("got %+v", ep)
	}
}

func TestParseEndpoint_UnsupportedSchemeRejectedByName(t *testing.T) {
	_, err := ParseEndpoint("ftp://server.example.com")
	if !errors.Is(err, ErrUnsupportedScheme) {
		t.Fatalf("expected ErrUnsupportedScheme, got %v", err)
	}
	if !strings.Contains(err.Error(), "ftp") {
		t.Fatalf("error should name the offending scheme: %v", err)
	}
	if !strings.Contains(err.Error(), "vectorizer") {
		t.Fatalf("error should list valid schemes: %v", err)
	}
}

func TestParseEndpoint_EmptyStringRejected(t *testing.T) {
	_, err := ParseEndpoint("")
	if !errors.Is(err, ErrEmptyURL) {
		t.Fatalf("expected ErrEmptyURL, got %v", err)
	}
	_, err = ParseEndpoint("   ")
	if !errors.Is(err, ErrEmptyURL) {
		t.Fatalf("expected ErrEmptyURL for whitespace, got %v", err)
	}
}

func TestParseEndpoint_UserinfoRejected(t *testing.T) {
	// Both schemes — credentials must go through HELLO, not the URL,
	// to avoid logging or shell-history-saving a token-bearing URL.
	_, err := ParseEndpoint("vectorizer://user:pass@host:15503")
	if !errors.Is(err, ErrCredentialsInURL) {
		t.Fatalf("vectorizer scheme: expected ErrCredentialsInURL, got %v", err)
	}
	_, err = ParseEndpoint("https://user:secret@api.example.com")
	if !errors.Is(err, ErrCredentialsInURL) {
		t.Fatalf("https scheme: expected ErrCredentialsInURL, got %v", err)
	}
}

func TestParseEndpoint_MalformedPortRejected(t *testing.T) {
	_, err := ParseEndpoint("vectorizer://host:not-a-port")
	if err == nil {
		t.Fatal("expected error for non-numeric port")
	}
	if !strings.Contains(err.Error(), "invalid port") {
		t.Fatalf("error should mention 'invalid port': %v", err)
	}
}

func TestParseEndpoint_IPv6WithPort(t *testing.T) {
	ep, err := ParseEndpoint("vectorizer://[::1]:15503")
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if ep.Kind != EndpointRPC || ep.Host != "[::1]" || ep.Port != 15503 {
		t.Fatalf("got %+v", ep)
	}
}

func TestParseEndpoint_IPv6WithoutPortDefaults(t *testing.T) {
	ep, err := ParseEndpoint("vectorizer://[::1]")
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if ep.Kind != EndpointRPC || ep.Host != "[::1]" || ep.Port != DefaultRPCPort {
		t.Fatalf("got %+v", ep)
	}
}

File: sdks/typescript/tests/rpc/endpoint.test.ts
/**
 * Unit tests for `parseEndpoint` — the canonical URL parser.
 *
 * The contract is shared with the Rust + Python SDKs; the golden URL
 * forms here are the same ones tested in `sdks/rust/src/rpc/endpoint.rs`
 * and `sdks/python/tests/rpc/test_endpoint.py`. Keeping them aligned
 * across SDKs prevents subtle behaviour drift between language
 * runtimes.
 */

import { describe, expect, test } from 'vitest';

import {
  DEFAULT_RPC_PORT,
  EndpointParseError,
  parseEndpoint,
} from '../../src/rpc/endpoint';

describe('parseEndpoint', () => {
  test('vectorizer:// with explicit host and port', () => {
    expect(parseEndpoint('vectorizer://example.com:9000')).toEqual({
      kind: 'rpc',
      host: 'example.com',
      port: 9000,
    });
  });

  test('vectorizer:// without port defaults to 15503', () => {
    expect(parseEndpoint('vectorizer://example.com')).toEqual({
      kind: 'rpc',
      host: 'example.com',
      port: DEFAULT_RPC_PORT,
    });
    expect(DEFAULT_RPC_PORT).toBe(15503);
  });

  test('bare host:port (no scheme) routes to RPC', () => {
    expect(parseEndpoint('localhost:15503')).toEqual({
      kind: 'rpc',
      host: 'localhost',
      port: 15503,
    });
  });

  test('http:// and https:// route to REST', () => {
    expect(parseEndpoint('http://localhost:15002')).toEqual({
      kind: 'rest',
      url: 'http://localhost:15002',
    });
    expect(parseEndpoint('https://api.example.com')).toEqual({
      kind: 'rest',
      url: 'https://api.example.com',
    });
  });

  test('unsupported scheme is rejected by name', () => {
    expect(() => parseEndpoint('ftp://server.example.com')).toThrow(
      EndpointParseError,
    );
    try {
      parseEndpoint('ftp://server.example.com');
    } catch (e) {
      expect(String(e)).toContain('ftp');
      expect(String(e)).toContain('vectorizer');
    }
  });

  test('empty string is rejected', () => {
    expect(() => parseEndpoint('')).toThrow(/empty/);
    expect(() => parseEndpoint('   ')).toThrow(/empty/);
  });

  test('userinfo credentials are rejected', () => {
    // Both schemes — credentials must go through HELLO, not the URL,
    // to avoid logging or shell-history-saving a token-bearing URL.
    expect(() => parseEndpoint('vectorizer://user:pass@host:15503')).toThrow(
      /credentials/,
    );
    expect(() => parseEndpoint('https://user:secret@api.example.com')).toThrow(
      /credentials/,
    );
  });

  test('malformed port is rejected', () => {
    expect(() => parseEndpoint('vectorizer://host:not-a-port')).toThrow(/invalid port/);
  });

  test('IPv6 literal with port', () => {
    expect(parseEndpoint('vectorizer://[::1]:15503')).toEqual({
      kind: 'rpc',
      host: '[::1]',
      port: 15503,
    });
  });

  test('IPv6 literal without port defaults', () => {
    expect(parseEndpoint('vectorizer://[::1]')).toEqual({
      kind: 'rpc',
      host: '[::1]',
      port: DEFAULT_RPC_PORT,
    });
  });

  test('non-string input is rejected', () => {
    // @ts-expect-error — intentionally violating the type to test runtime guard.
    expect(() => parseEndpoint(15503)).toThrow(/must be a string/);
  });
});
